cmd/util/ledger/migrations/fix_broken_data_migration.go

```diff
@@ -0,0 +1,191 @@
+package migrations
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/atree"
+
+	"github.com/onflow/cadence/runtime/common"
+
+	"github.com/onflow/flow-go/cmd/util/ledger/reporters"
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/ledger/common/convert"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+type FixSlabsWithBrokenReferencesMigration struct {
+	log           zerolog.Logger
+	rw            reporters.ReportWriter
+	accountsToFix map[common.Address]string
+	nWorkers      int
+}
+
+var _ AccountBasedMigration = &FixSlabsWithBrokenReferencesMigration{}
+
+const fixSlabsWithBrokenReferencesName = "fix-slabs-with-broken-references"
+
+func NewFixBrokenReferencesInSlabsMigration(
+	rwf reporters.ReportWriterFactory,
+	accountsToFix map[common.Address]string,
+) *FixSlabsWithBrokenReferencesMigration {
+	return &FixSlabsWithBrokenReferencesMigration{
+		rw:            rwf.ReportWriter(fixSlabsWithBrokenReferencesName),
+		accountsToFix: accountsToFix,
+	}
+}
+
+func (m *FixSlabsWithBrokenReferencesMigration) InitMigration(
+	log zerolog.Logger,
+	_ []*ledger.Payload,
+	nWorkers int,
+) error {
+	m.log = log.
+		With().
+		Str("migration", fixSlabsWithBrokenReferencesName).
+		Logger()
+	m.nWorkers = nWorkers
+
+	return nil
+}
+
+func (m *FixSlabsWithBrokenReferencesMigration) MigrateAccount(
+	_ context.Context,
+	address common.Address,
+	oldPayloads []*ledger.Payload,
+) (
+	newPayloads []*ledger.Payload,
+	err error,
+) {
+
+	if _, exist := m.accountsToFix[address]; !exist {
+		return oldPayloads, nil
+	}
+
+	migrationRuntime, err := NewAtreeRegisterMigratorRuntime(address, oldPayloads)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create cadence runtime: %w", err)
+	}
+
+	storage := migrationRuntime.Storage
+
+	// Load all atree registers in storage
+	for _, payload := range oldPayloads {
+		registerID, _, err := convert.PayloadToRegister(payload)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert payload to register: %w", err)
+		}
+
+		if !registerID.IsSlabIndex() {
+			continue
+		}
+
+		// Convert the register ID to a storage ID.
+		slabID := atree.NewStorageID(
+			atree.Address([]byte(registerID.Owner)),
+			atree.StorageIndex([]byte(registerID.Key[1:])))
+
+		// Retrieve the slab.
+		_, _, err = storage.Retrieve(slabID)
+		if err != nil {
+			return nil, fmt.Errorf("failed to retrieve slab %s: %w", slabID, err)
+		}
+	}
```

Comment:
The slab loading logic already exists in `checkStorageHealth`. Maybe factor it out of there into a function, e.g. `loadPayloadSlabs` and use it in `checkStorageHealth` and here

Reply:
Done in 64fc18889cea51ff696037f8c5af2e043f8c2fac.